smalltalksrc/VMMakerTests/VMSpurEphemeronsAlgorithmTest.class.st

```diff
@@ -0,0 +1,361 @@
+	ephemeronObjectOopTwo := memory remapObj: ephemeronObjectOopTwo."
+	
+	self assert: memory dequeueMourner equals: ephemeronObjectOopTwo.
+	self assert: memory dequeueMourner equals: ephemeronObjectOopOne
+]
+
+{ #category : #'tests - finalization - algorithm' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsOldObjectsYoung [
+	
+	ephemeronObjectOopOne := self newEphemeronObjectOldSpace.
+	ephemeronObjectOopTwo := self newEphemeronObjectOldSpace.
+	nonEphemeronObjectOopOne := self newObjectWithSlots: 1.
+	nonEphemeronObjectOopTwo := self newZeroSizedObject.
+	"Make object 1 reference object 2"
+	memory storePointer: 0 ofObject: nonEphemeronObjectOopOne withValue: nonEphemeronObjectOopTwo. 
+
+	self addKeysToEphemerons.
+	self keepEphemeronsInVMVariables.
+	
+	"Collect both non ephemeron objects"
+	memory doScavenge:1 "Tenured by age".
+	memory fullGC.
+
+	"ephemeronObjectOopOne := memory remapObj: ephemeronObjectOopOne.
+	ephemeronObjectOopTwo := memory remapObj: ephemeronObjectOopTwo."
+	
+	self assert: memory dequeueMourner equals: ephemeronObjectOopTwo.
+	self assert: memory dequeueMourner equals: ephemeronObjectOopOne
+]
+
+{ #category : #'tests - finalization - algorithm' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsOldObjectsYoungInverse [
+
+	ephemeronObjectOopOne := self newEphemeronObjectOldSpace.
+	ephemeronObjectOopTwo := self newEphemeronObjectOldSpace.
+	nonEphemeronObjectOopOne := self newZeroSizedObject.
+	nonEphemeronObjectOopTwo := self newObjectWithSlots: 1.
+	"Make object 2 reference object 1"
+	memory storePointer: 0 ofObject: nonEphemeronObjectOopTwo withValue: nonEphemeronObjectOopOne. 
+
+	self addKeysToEphemerons.
+	self keepEphemeronsInVMVariables.
+	
+	"Collect both non ephemeron objects"
+	memory doScavenge:1 "Tenured by age".
+	memory fullGC.
+
+	"ephemeronObjectOopOne := memory remapObj: ephemeronObjectOopOne.
+	ephemeronObjectOopTwo := memory remapObj: ephemeronObjectOopTwo."
+	
+	self assert: memory dequeueMourner equals: ephemeronObjectOopTwo.
+	self assert: memory dequeueMourner equals: ephemeronObjectOopOne
+]
+
+{ #category : #'tests - ephemerons with same key' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsWithSameKeyNewSpace [
+
+	ephemeronObjectOopOne := self newEphemeronObject.
+	ephemeronObjectOopTwo := self newEphemeronObject.
+	nonEphemeronObjectOopOne := self newZeroSizedObject.
+
+	"Both ephemerons share the same key"
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopOne
+		withValue: nonEphemeronObjectOopOne.
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopTwo
+		withValue: nonEphemeronObjectOopOne.
+
+	self keepObjectInVMVariable1: ephemeronObjectOopOne.
+	self keepObjectInVMVariable2: ephemeronObjectOopTwo.
+
+	"Collect ephemerons"
+	memory doScavenge: 1.
+
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner).
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner)
+]
```

Comment:
This test will fail so far by design.

Ephemerons are processed one-by-one during the scavenge.
When the first ephemeron is processed, the key is marked to survive (copied), and suddenly the second ephemeron's key survived and should not be copied.

There is a test, with the same setup contradicting your example

`testScavengeTwoEphemeronObjectsReferencingSameDyingObjectsShouldAddOnlyOneEphemeron`

This will force only one ephemeron to be finalized per cycle, which is indeed not optimal.
But it should be worked separately.